docs/appengine/firebase/tictactoe/main.go

```diff
@@ -0,0 +1,295 @@
+package tictactoe
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/zabawaba99/firego"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/datastore"
+	"google.golang.org/appengine/log"
+	"google.golang.org/appengine/user"
+)
+
+func init() {
+	http.HandleFunc("/", indexHandler)
+	http.HandleFunc("/opened", openedHandler)
+	http.HandleFunc("/move", moveHandler)
+	http.HandleFunc("/delete", deleteHandler)
+}
+
+var tmpl = template.Must(template.ParseFiles("template.html"))
+
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+
+	config, err := ioutil.ReadFile("firebase_config.html")
+	if err != nil {
+		handleError(w, r, "Could not read config file", err)
+		return
+	}
+
+	u := user.Current(ctx)
+
+	var game *Game
+	if gameID := r.FormValue("g"); gameID == "" {
+		game = NewGame()
+		game.K = datastore.NewKey(ctx, "Game", u.ID, 0, nil)
+		game.UserX = u.ID
+
+		if _, err := datastore.Put(ctx, game.K, game); err != nil {
+			handleError(w, r, "Could not start game", err)
+			return
+		}
+	} else {
+		// Existing game, join it.
+		if err := datastore.RunInTransaction(ctx, func(ctx context.Context) error {
+			g, err := gameFromRequest(r)
+			if err != nil {
+				return err
+			}
+			game = g
+			if game.UserO == "" {
+				game.UserO = u.ID
+				_, err := datastore.Put(ctx, game.K, game)
+				return err
+			}
+			return nil
+		}, nil); err != nil {
+			handleError(w, r, "Could not join game", err)
+			return
+		}
+	}
+
+	gameKey := game.K.Encode()
+	channelID := u.ID + gameKey
+
+	tok, err := createToken(ctx, channelID)
+	if err != nil {
+		handleError(w, r, "Could not create auth token", err)
+		return
+	}
+
+	d := struct {
+		GameKey        string
+		GameLink       string
+		Me             string
+		Token          string
+		ChannelID      string
+		FirebaseConfig template.HTML
+		InitialMessage *Game
+	}{
+		FirebaseConfig: template.HTML(config),
+		Token:          tok,
+		GameKey:        gameKey,
+		GameLink:       r.URL.Host + "/?g=" + gameKey,
+		Me:             u.ID,
+		ChannelID:      channelID,
+		InitialMessage: game,
+	}
+	if err := tmpl.Execute(w, d); err != nil {
+		handleError(w, r, "Could not execute template", err)
+		return
+	}
+}
+
+func createToken(ctx context.Context, channelID string) (string, error) {
```

Comment:
Does Go have a jwt library that can use App Engine to sign bytes? If so, why not use that?